qiskit_machine_learning/kernels/statevector_kernel.py

```diff
@@ -0,0 +1,129 @@
+    This stash can be cleared using :meth:`clear_cache`. By default, the cache is cleared
+    when :meth:`evaluate` is called.
+
+    """
+
+    def __init__(
+        self, *, feature_map: QuantumCircuit | None = None, statevector_type: Type[SV] = Statevector
+    ) -> None:
+        """
+        Args:
+            feature_map: Parameterized circuit to be used as the feature map. If ``None`` is given,
+                :class:`~qiskit.circuit.library.ZZFeatureMap` is used with two qubits. If there's
+                a mismatch in the number of qubits of the feature map and the number of features
+                in the dataset, then the kernel will try to adjust the feature map to reflect the
+                number of features.
+            statevector_type: The type of Statevector that will be instantiated using the
+                ``feature_map`` quantum circuit and used to compute the fidelity kernel. This type
+                should inherit from and defaults to :class:`~qiskit.quantum_info.Statevector`.
+        """
+        super().__init__(feature_map=feature_map)
+
+        self._statevector_type = statevector_type
+        self._statevector_cache: dict[tuple[float, ...], np.ndarray] = {}
+
+    def evaluate(
+        self, x_vec: np.ndarray, y_vec: np.ndarray | None = None, clear_cache: bool = True
+    ) -> np.ndarray:
+        r"""
+        Construct kernel matrix for given data.
+
+        If ``y_vec`` is ``None``, self inner product is calculated.
+
+        Args:
+            x_vec: 1D or 2D array of datapoints, :math:`N\times D`, where :math:`N` is the number of
+                datapoints, :math:`D` is the feature dimension.
+            y_vec: 1D or 2D array of datapoints, :math:`M\times D`, where :math:`M` is the number of
+                datapoints, :math:`D` is the feature dimension.
+            clear_cache: Boolean that determines whether the statevector cache is cleared when
+                evaluate is called. Defaults to ``True``.
+
+        Returns:
+            2D matrix, :math:`N\times M`.
+        """
+        if clear_cache:
+            self.clear_cache()
+
+        x_vec, y_vec = self._validate_input(x_vec, y_vec)
+
+        if y_vec is None:
+            y_vec = x_vec
+
+        kernel_shape = (x_vec.shape[0], y_vec.shape[0])
+
+        x_svs = [self._get_statevector(x) for x in x_vec]
+        y_svs = [self._get_statevector(y) for y in y_vec]
+
+        kernel_matrix = np.ones(kernel_shape)
+        for i, x in enumerate(x_svs):
+            for j, y in enumerate(y_svs):
+                if np.array_equal(x, y):
+                    continue
+                kernel_matrix[i, j] = self._compute_kernel_element(x, y)
+
+        return kernel_matrix
+
+    @staticmethod
+    def _compute_kernel_element(x: np.ndarray, y: np.ndarray) -> float:
+        return np.abs(np.conj(x) @ y) ** 2
+
+    def _get_statevector(self, param_values) -> np.ndarray:
```

Comment:
I've refactored the code to use `functools.lru_cache(maxsize=None)` as a Python3.7- and 3.8-compatible `functools.cache`. Since we're wrapping an instance method here, I've made the cache container local to the instance so it can be cleaned up properly, rather than use a decorator.

Reply:
It probably makes sense to put a limit on the cache size, but I'm not sure what a typical user would require. @ElePT do you have an idea of cache size requirements based on what the client that required the protocol implementation was using?

Reply:
Given how you have done things you could have say a `cache_size` parameter on the constructor so whatever default you end up with a user can still adjust it via that parameter.

Reply:
Out of curiosity do you know how much difference the caching makes to the performance?

Reply:
I can try to test on Pegasos, but honestly, I don't expect dramatic improvements.